backend/server/src/services/tournamentService.ts

```diff
@@ -67,4 +100,262 @@ export class TournamentService {
 
     return await tournament.toObject();
   }
+
+  // update - add custom match
+  public async addMatchToTournament(
+    tournamentId: string,
+    unsavedMatch: UnsavedMatch
+  ): Promise<Tournament> {
+    // custom add match to tournament
+    const tournament = await TournamentModel.findById(tournamentId).exec();
+    if (tournament === null || tournament === undefined) {
+      throw new NotFoundError({
+        message: "Tournament not found"
+      });
+    }
+
+    // validation
+    for (let i = 0; i < unsavedMatch.players.length; i += 1) {
+      if (!tournament.players.includes(unsavedMatch.players[i].id)) {
+        const user = await UserModel.findById(
+          unsavedMatch.players[i].id
+        ).exec();
+
+        if (user === null || user === undefined) {
+          throw new NotFoundError({
+            message: "Player not found!"
+          });
+        }
+        throw new BadRequestError({
+          message: `Cannot create the match: ${user.firstName} ${user.lastName} is not registered for this tournament.`
+        });
+      }
+    }
+
+    const newMatch = await MatchModel.create(unsavedMatch);
+    tournament.matchSchedule.push(newMatch._id);
+    await tournament.save();
+    return await tournament.toObject();
+  }
+
+  // update - auto generate match
+  public async generateTournamentSchedule(
+    tournamentId: string
+  ): Promise<Tournament> {
+    const tournament = await TournamentModel.findById(tournamentId).exec();
+
+    if (tournament === null || tournament === undefined) {
+      throw new NotFoundError({
+        message: "Tournament not found"
+      });
+    }
+
+    if (tournament.players.length !== tournament.maxPlayers) {
+      throw new BadRequestError({
+        message: "Tournament player count has not reached the maximum capacity"
+      });
+    }
+
+    let matches: UnsavedMatch[] = [];
+
+    switch (tournament.tournamentType) {
+      case TournamentType.RoundRobin:
+        if (
+          this.calculateRoundRobinMatches(tournament.maxPlayers) ===
+          tournament.matchSchedule.length
+        ) {
+          throw new BadRequestError({
+            message:
+              "We have already created the match schedule for this tournament."
+          });
+        }
+        matches = this.generateRoundRobinSchedule(tournament.players);
+        break;
+      case TournamentType.Playoff:
+        matches = await this.generatePlayoffSchedule(
+          tournament.players,
+          tournament.matchSchedule
+        );
+        break;
```

Comment:
This could be split to a smaller function for readability and maintainability sake

Reply:
This function has been reduced to much smaller in size. It doesn't need to be broken down anymore. most of the conditions have been changed as well.